### ISSUE TITLE
Questions about calculating the gradient

### ISSUE THREAD

**Issue by handsomelys** [maple-research-lab/CaCo#2]
Sorry to bother you, I wanted to ask for a problem in the code.
` g1 = torch.einsum('cn,nk->ck', [q_pred.T, p_qd1]) / logits1.shape[0] - torch.mul(torch. Mean(torch.mul(p_qd1, logits_keep1), dim=0), d_norm1)`
As the code showed above, in my understanding, this part of code:
`torch.einsum('cn,nk->ck', [q_pred.T, p_qd1]) / logits1.shape[0]`
The gradient has been calculated (according to Eq.7)
So what is the latter calculation for?
`- torch.mul(torch. Mean(torch.mul(p_qd1, logits_keep1), dim=0), d_norm1)`
Or is there a problem with my understanding, I tried to use CaCo for my task, but this part is a little incomprehensible.
Looking forward to your reply!

**Comment by wang3702** [comment]
please refer our discussion in AdCo: https://github.com/maple-research-lab/AdCo/issues/8. I think this can help you understand why g1 is in this form.

**Comment by handsomelys** [comment]
Thanks for your reply. But I still can't understand. Can you please explain this to me further?
Are there any equations in CaCo's paper about this question?
Or should I refer to the AdCo's paper?

**Comment by wang3702** [comment]
Please see Eq.(4) in caco paper, if b_j is not positive for any one of given queries, then the update rule of b_j is exactly the same as AdCo. 
Therefore, checking code here https://github.com/maple-research-lab/CaCo/blob/main/training/train_caco.py#L193-L194, after resetting the p(b_j|z) for those positive counterparts, then it's the same update rule as AdCo. 
Here you think https://github.com/maple-research-lab/CaCo/blob/main/training/train_caco.py#L196 is so complicated, because you need to update the normalized vector, while z in Eq.(4) refers to the normalized the vector. The real gradient needs to be applied to the unnormalized vector, which is discussed in https://github.com/maple-research-lab/AdCo/issues/8. That's why it needs to be written in 
```
 g1 = torch.einsum('cn,nk->ck', [q_pred.T, p_qd1]) / logits1.shape[0] - torch.mul(torch. Mean(torch.mul(p_qd1, logits_keep1), dim=0), d_norm1)
```

**Comment by handsomelys** [comment]
Thanks for your reply! I feel like I kind of get it.
But doesn't this code [https://github.com/maple-research-lab/CaCo/blob/main/model/CaCo.py#L128](url) mean that negative samples have been normalized before doing einsum?
And in the discussion in [https://github.com/maple-research-lab/AdCo/issues/8](url), what is this term means? What is the explicit formula for S_qinu? I can not find this formula in the AdCo's paper. Does it is the same as the equation(8) in CaCo's paper?
![image](https://github.com/maple-research-lab/CaCo/assets/48092352/1ce9a210-474c-4402-9910-b15d9cc993d4)

**Comment by handsomelys** [comment]
I think I figured it out, thank you very much for your kind answer!

**Comment by wang3702** [comment]
No problem.
